libra-node/src/main_node.rs

```diff
@@ -255,60 +255,68 @@ pub fn setup_environment(node_config: &mut NodeConfig) -> LibraHandle {
+        network_configs.push((RoleType::Validator, network_config));
     }
 
-    for mut full_node_network in node_config.full_node_networks.iter_mut() {
-        let (runtime, mut network_builder) = setup_network(
-            &mut full_node_network,
-            RoleType::FullNode,
-            Arc::clone(&db_rw.reader),
-            waypoint,
-        );
+    let mut validator_count = 0;
+    // Instantiate every network and collect the requisite endpoints for state_sync, mempool, and consensus.
+    for (role, network_config) in network_configs {
+        // Perform common instantiation steps
+        let (runtime, mut network_builder) =
+            setup_network(network_config, role, Arc::clone(&db_rw.reader), waypoint);
         let peer_id = network_builder.peer_id();
 
-        network_runtimes.push(runtime);
+        // Create the endpoints to connect the Network to StateSynchronizer.
         let (state_sync_sender, state_sync_events) =
             state_synchronizer::network::add_to_network(&mut network_builder);
         state_sync_network_handles.push((peer_id, state_sync_sender, state_sync_events));
+
+        // Create the endpoints to connect the network to MemPool.
         let (mempool_sender, mempool_events) = libra_mempool::network::add_to_network(
             &mut network_builder,
             node_config.mempool.max_broadcasts_per_peer,
         );
         mempool_network_handles.push((peer_id, mempool_sender, mempool_events));
 
-        // Start the network provider.
+        match role {
+            // Perform steps relevant specifically to Validator networks.
+            RoleType::Validator => {
+                validator_count += 1;
+                // A valid config is allowed to have exactly one ValidatorNetwork
+                assert_eq!(1, validator_count);
+                // Set up to listen for network configuration changes from StateSync.
+                // TODO:  move this inside network_builder.
+                if let Some(conn_mgr_reqs_tx) = network_builder.conn_mgr_reqs_tx() {
+                    let (simple_discovery_reconfig_subscription, simple_discovery_reconfig_rx) =
+                        gen_simple_discovery_reconfig_subscription();
+                    reconfig_subscriptions.push(simple_discovery_reconfig_subscription);
+                    let network_config_listener =
+                        ConfigurationChangeListener::new(conn_mgr_reqs_tx, RoleType::Validator);
+                    runtime
+                        .handle()
+                        .spawn(network_config_listener.start(simple_discovery_reconfig_rx));
+                };
+
+                consensus_network_handles = Some(consensus::network_interface::add_to_network(
+                    &mut network_builder,
+                ));
+            }
+            // Currently no FullNode network specific steps.
+            RoleType::FullNode => (),
```

Comment:
then we don't have to have this awkward situation

Reply:
I actually see this as a feature not a bug for the time being --- similar to not wanting to overthink moving things into setup_network at this point, keeping this specific hook in place while refactoring the builders will be useful (and also makes it very explicit that there is no functionality that is currently full node specific/where that functionality should be configured if it appears)